service/history/replicationTaskProcessor.go

```diff
@@ -202,40 +205,38 @@ Loop:
 			}
 
 			scope.UpdateGauge(metrics.LastRetrievedMessageID, float64(p.lastRetrievedMessageID))
-			p.noTaskBackoffRetrier.Reset()
+			p.noTaskRetrier.Reset()
 		case <-p.done:
+			p.logger.Info("ReplicationTaskProcessor shutting down.")
 			return
 		}
 	}
 }
 
-func (p *ReplicationTaskProcessor) processTask(replicationTask *r.ReplicationTask) {
-	var err error
+func (p *ReplicationTaskProcessor) processTask(replicationTask *r.ReplicationTask) error {
+	err := backoff.Retry(func() error {
+		return p.processTaskOnce(replicationTask)
+	}, p.taskRetryPolicy, isTransientRetryableError)
 
-	for execute := true; execute; execute = err != nil {
-		err = backoff.Retry(func() error {
-			return p.processTaskOnce(replicationTask)
-		}, p.retryPolicy, isTransientRetryableError)
+	if err != nil {
+		p.logger.Error(
+			"Failed to apply replication task after retry. Putting task into DLQ.",
+			tag.TaskID(replicationTask.GetSourceTaskId()),
+			tag.Error(err),
+		)
 
-		if err != nil {
-			// TODO: insert into our own dlq in cadence persistence?
-			// p.nackMsg(msg, err, logger)
-			p.logger.Error(
-				"Failed to apply replication task after retry.",
-				tag.TaskID(replicationTask.GetSourceTaskId()),
-				tag.Error(err),
-			)
-		}
+		return p.putReplicationTaskToDLQ(replicationTask)
 	}
+
+	return nil
 }
 
 func (p *ReplicationTaskProcessor) processTaskOnce(replicationTask *r.ReplicationTask) error {
 	var err error
 	var scope int
 	switch replicationTask.GetTaskType() {
 	case r.ReplicationTaskTypeDomain:
-		scope = metrics.DomainReplicationTaskScope
-		err = p.handleDomainReplicationTask(replicationTask)
+		// Domain replication task should be handled in worker (domainReplicationMessageProcessor)
```

Comment:
if this case is not expected, either panic or return an error

Reply:
Done